### PR TITLE
Cooler zoomify

### PR DESCRIPTION
**Pull request by jianhong**
## PR checklist

- [ x ] If you've added a new tool - have you followed the module conventions in the [contribution docs](https://github.com/nf-core/modules/tree/master/.github/CONTRIBUTING.md)
- [ x ] If necessary, include test data in your PR.
- [ x ] Remove all TODO statements.
- [ x ] Emit the `<SOFTWARE>.version.txt` file.
- [ x ] Follow the naming conventions.
- [ x ] Follow the parameters requirements.
- [ x ] Follow the input/output options guidelines.
- [ x ] Add a resource `label`
- [ x ] Use BioConda and BioContainers if possible to fulfil software requirements.
- Ensure that the test works with either Docker / Singularity. Conda CI tests can be quite flaky:
    - [ x ] `PROFILE=docker pytest --tag <MODULE> --symlink --keep-workflow-wd`
    - [ ] `PROFILE=singularity pytest --tag <MODULE> --symlink --keep-workflow-wd`
    - [ x ] `PROFILE=conda pytest --tag <MODULE> --symlink --keep-workflow-wd`

### CODE REVIEW

File: software/cooler/zoomify/functions.nf [hunk 1]
```diff
@@ -0,0 +1,70 @@
+/*
```
Comment (drpatelh) [inline software/cooler/zoomify/functions.nf:1]: This is the old `functions.nf` file. Be great if you can update this for all the current PRs.

File: software/cooler/zoomify/main.nf [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+
+    script:
+    def software = getSoftwareName(task.process)
+    def prefix   = options.suffix ? "${meta.id}${options.suffix}" : "${meta.id}"
+    def cool_bin = "${meta.bin}"
+    def res      = options.args2?:""
+    """
+    cooler zoomify \\
+        $options.args \\
+        -r ${cool_bin}$res \\
```
Comment (drpatelh) [inline software/cooler/zoomify/main.nf:36]: This looks like an optional argument? so should be passed to the module via `options.args`

File: software/cooler/zoomify/main.nf [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+        -n $task.cpus \\
+        -o ${prefix}.${cool_bin}.mcool \\
+        ${cool}
```
Comment (drpatelh) [inline software/cooler/zoomify/main.nf:39]:
```suggestion
        -o ${prefix}.mcool \\
        $cool
```

File: software/cooler/zoomify/main.nf (hunk 2 above)
Comment (jianhong) [reply software/cooler/zoomify/main.nf:36]: This parameter is related to meta data so I keep it separately. Do that work?

File: software/cooler/zoomify/main.nf [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+
+    output:
+    tuple val(meta), path("*.mcool"), emit: mcool
+    path "*.version.txt"            , emit: version
+
+    script:
+    def software = getSoftwareName(task.process)
+    def prefix   = options.suffix ? "${meta.id}${options.suffix}" : "${meta.id}"
+    def cool_bin = "${meta.bin}"
+    def res      = options.args2?:""
```
Comment (drpatelh) [inline software/cooler/zoomify/main.nf:32]: Hmmm...what is this doing? `options.args2` is reserved for instances where you need to run 2 independent tools in the same module.

Comment (drpatelh) [inline software/cooler/zoomify/main.nf:32]:
```suggestion
```

File: software/cooler/zoomify/main.nf (hunk 2 above)
Comment (drpatelh) [inline software/cooler/zoomify/main.nf:36]:
```suggestion
```

File: tests/software/cooler/zoomify/main.nf [hunk 5]
```diff
@@ -0,0 +1,12 @@
+#!/usr/bin/env nextflow
+
+nextflow.enable.dsl = 2
+
+include { COOLER_ZOOMIFY } from '../../../../software/cooler/zoomify/main.nf' addParams( options: [:] )
```
Comment (drpatelh) [inline tests/software/cooler/zoomify/main.nf:5]:
```suggestion
include { COOLER_ZOOMIFY } from '../../../../software/cooler/zoomify/main.nf' addParams( options: ['args':'-r 2,4,8'] )
```

File: tests/software/cooler/zoomify/main.nf [hunk 6]
```diff
@@ -0,0 +1,12 @@
...
+
+workflow test_cooler_zoomify {
+   input = [ [ id:'test', bin:"2,4,8" ], // meta map
```
Comment (drpatelh) [inline tests/software/cooler/zoomify/main.nf:8]:
```suggestion
   input = [ [ id:'test' ], // meta map
```

File: tests/software/cooler/zoomify/test.yml [hunk 7]
```diff
@@ -0,0 +1,7 @@
+- name: cooler zoomify test_cooler_zoomify
+  command: nextflow run tests/software/cooler/zoomify -entry test_cooler_zoomify -c tests/config/nextflow.config
+  tags:
+    - cooler/zoomify
+    - cooler
+  files:
+    - path: output/cooler/test.2,4,8.mcool
```
Comment (drpatelh) [inline tests/software/cooler/zoomify/test.yml:7]:
```suggestion
    - path: output/cooler/test.mcool
```

File: software/cooler/zoomify/main.nf (hunk 2 above)
Comment (drpatelh) [reply software/cooler/zoomify/main.nf]: `meta` should typically contain sample information. I would have thought that `bins` isn't sample specific?

**Comment by drpatelh** [comment]
Ok. I have pushed some commits here to hopefully give you an idea as to how the module should be organised. Sorry, I kind of just updated things without asking 😅 But feel free to make counter comments if I have done the wrong thing. Just need to pull in my changes and update the `functions.nf` file now.

I will comment on the others for you to make the changes but I guess you realise now why it's easier to have separate PRs for each module.

**Comment by jianhong** [comment]
> Ok. I have pushed some commits here to hopefully give you an idea as to how the module should be organised. Sorry, I kind of just updated things without asking 😅 But feel free to make counter comments if I have done the wrong thing. Just need to pull in my changes and update the `functions.nf` file now.
> 
> I will comment on the others for you to make the changes but I guess you realise now why it's easier to have separate PRs for each module.

NO problem. I will go through the modification later. Yes I learned a lot during this process. Thank you.

**Comment by jianhong** [comment]
@drpatelh I updated the functions.nf to newest version. I have a question about the change of the main.nf. What is the best way to append arguments to options.args for each sample? For example the default setting of options.arg = "", and I want to add "-r 5000N" to one sample and "-r 10000N" to another sample?

**Comment by jianhong** [comment]
updated the module to fit new template.

File: tests/modules/cooler/zoomify/test.yml [hunk 8]
```diff
@@ -0,0 +1,7 @@
+- name: cooler zoomify test_cooler_zoomify
+  command: nextflow run tests/modules/cooler/zoomify -entry test_cooler_zoomify -c tests/config/nextflow.config
+  tags:
+    - cooler
+    - cooler/zoomify
+  files:
+    - path: output/cooler/test.mcool
```
Comment (grst) [inline tests/modules/cooler/zoomify/test.yml:7]: Can you please add an md5sum, or, if it is not stable between runs, some keyword(s) that should be contained in the file?

**Review by grst — COMMENTED** [review]
Looks good to me, just a missing checksum in the `test.yml`.

File: tests/modules/cooler/zoomify/test.yml (hunk 8 above)
Comment (jianhong) [reply tests/modules/cooler/zoomify/test.yml:7]: Yes it is not stable between runs. I will add some keyword(s) to validate the output.

**Review by grst — APPROVED** [review]
LGTM!

File: modules/cooler/zoomify/functions.nf [hunk 9]
```diff
@@ -0,0 +1,68 @@
+//
+//  Utility functions used in nf-core DSL2 module files
+//
+
+//
```
Comment (drpatelh) [inline modules/cooler/zoomify/functions.nf:5]: Need to update this file?

File: modules/cooler/zoomify/main.nf [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+    } else {
+        container "quay.io/biocontainers/cooler:0.8.11--pyh3252c3a_0"
+    }
+
+    input:
+    tuple val(meta), path(cool)
+
+    output:
+    tuple val(meta), path("*.mcool"), emit: mcool
+    path "*.version.txt"            , emit: version
```
Comment (drpatelh) [inline modules/cooler/zoomify/main.nf:26]: And to use the latest versions syntax.

File: modules/cooler/dump/main.nf [hunk 11]
```diff
@@ -20,18 +20,20 @@ process COOLER_DUMP {
 
     input:
     tuple val(meta), path(cool)
+    val resolution
 
     output:
     tuple val(meta), path("*.bedpe"), emit: bedpe
     path "versions.yml"                  , emit: versions
```
Comment (drpatelh) [inline modules/cooler/dump/main.nf:27]:
```suggestion
    tuple val(meta), path("*.bedpe"), emit: bedpe
    path "versions.yml"             , emit: versions
```

File: modules/cooler/zoomify/main.nf (hunk 10 above)
Comment (drpatelh) [inline modules/cooler/zoomify/main.nf:26]:
```suggestion
    tuple val(meta), path("*.mcool"), emit: mcool
    path "versions.yml"             , emit: versions
```

File: modules/cooler/zoomify/main.nf [hunk 12]
```diff
@@ -0,0 +1,40 @@
...
+
+    script:
+    def software = getSoftwareName(task.process)
```
Comment (drpatelh) [inline modules/cooler/zoomify/main.nf:29]:
```suggestion
```

File: modules/cooler/zoomify/main.nf [hunk 13]
```diff
@@ -0,0 +1,40 @@
...
+    def prefix   = options.suffix ? "${meta.id}${options.suffix}" : "${meta.id}"
+    """
+    cooler zoomify \\
+        $options.args \\
+        -n $task.cpus \\
+        -o ${prefix}.mcool \\
+        $cool
+
+    echo \$(cooler --version 2>&1) | sed 's/cooler, version //' > ${software}.version.txt
```
Comment (drpatelh) [inline modules/cooler/zoomify/main.nf:38]: Update the syntax for this version command to latest.

File: modules/cooler/zoomify/meta.yml [hunk 14]
```diff
@@ -0,0 +1,41 @@
...
+output:
+  - meta:
+      type: map
+      description: |
+        Groovy Map containing sample information
+        e.g. [ id:'test', single_end:false ]
+  - version:
+      type: file
+      description: File containing software version
+      pattern: "*.{version.txt}"
```
Comment (drpatelh) [inline modules/cooler/zoomify/meta.yml:34]:
```suggestion
  - versions:
      type: file
      description: File containing software versions
      pattern: "versions.yml"
```

Comment (jianhong) [reply modules/cooler/zoomify/meta.yml]: I updated the version file.

File: tests/modules/cooler/dump/main.nf [hunk 15]
```diff
@@ -9,5 +9,5 @@ workflow test_cooler_dump {
     input = [ [ id:'test' ], // meta map
               file("https://raw.githubusercontent.com/open2c/cooler/master/tests/data/toy.asymm.16.cool", checkIfExists: true) ]
 
-    COOLER_DUMP ( input )
+    COOLER_DUMP ( input, [:] )
```
Comment (grst) [inline tests/modules/cooler/dump/main.nf:12]: How is this change related to the `zoomify` module?

File: tests/config/pytest_software.yml [hunk 16]
```diff
@@ -0,0 +1,712 @@
+abacas:
```
Comment (grst) [inline tests/config/pytest_software.yml:1]: This file is now called `pytest_modules.yml`. Please remove this file and update teh `pytest_modules` accordingly.

**Review by grst — CHANGES_REQUESTED** [review]

File: tests/modules/cooler/dump/main.nf (hunk 15 above)
Comment (jianhong) [reply tests/modules/cooler/dump/main.nf:12]: I updated cooler/dump also. This code is not related with `zoomify`. But because in the test I need to use cooler/dump to check the output, this should be also change.

File: tests/config/pytest_software.yml (hunk 16 above)
Comment (jianhong) [reply tests/config/pytest_software.yml]: I will update this.

File: modules/cooler/zoomify/main.nf [hunk 17]
```diff
@@ -0,0 +1,42 @@
...
+    } else {
+        container "quay.io/biocontainers/cooler:0.8.11--pyh3252c3a_0"
+    }
+
+    input:
+    tuple val(meta), path(cool)
+
+    output:
+    tuple val(meta), path("*.mcool"), emit: mcool
+    path "*.version.yml"            , emit: version
```
[the remark this answers is not in the document]
Comment (jianhong) [reply modules/cooler/zoomify/main.nf]: sorry for this mistake.

File: modules/cooler/dump/main.nf [hunk 18]
```diff
@@ -20,18 +20,20 @@ process COOLER_DUMP {
...
+    val resolution
 
     output:
     tuple val(meta), path("*.bedpe"), emit: bedpe
-    path "versions.yml"                  , emit: versions
+    path "versions.yml"             , emit: versions
 
     script:
     def prefix   = options.suffix ? "${meta.id}${options.suffix}" : "${meta.id}"
+    def postfix  = resolution?"::$resolution":""
```
Comment (FriederikeHanssen) [inline modules/cooler/dump/main.nf:31]: should be `suffix` . CAn you also align this properly with the line above?

File: modules/cooler/dump/main.nf [hunk 19]
```diff
@@ -20,18 +20,20 @@ process COOLER_DUMP {
...
     """
     cooler dump \\
         $options.args \\
         -o ${prefix}.bedpe \\
-        $cool
+        $cool$postfix
```
Comment (FriederikeHanssen) [inline modules/cooler/dump/main.nf:36]:
```suggestion
        $cool$suffix
```

File: modules/cooler/zoomify/functions.nf (hunk 9 above)
Comment (FriederikeHanssen) [reply modules/cooler/zoomify/functions.nf:5]: linting is passing now, so I guess this should be fine, right?

File: tests/modules/cooler/dump/main.nf (hunk 15 above)
Comment (FriederikeHanssen) [reply tests/modules/cooler/dump/main.nf:12]: I suppose this is resolved then?

**Comment by FriederikeHanssen** [comment]
@grst any open change requests? it think there are all adressed, but I think you need to confirm

File: modules/cooler/zoomify/functions.nf (hunk 9 above)
Comment (grst) [reply modules/cooler/zoomify/functions.nf:5]: yeah, that's the latest.

**Review by grst — APPROVED** [review]
Looks good now, thanks @FriederikeHanssen  and @jianhong!
